website/blog/2023-01-02-presto-hudi-deep-dive.md

```diff
@@ -0,0 +1,110 @@
+
+<!--truncate-->
+
+## Input formats and Record readers
+Before actually diving into the presto-hudi integration, it is important to go over a few involved concepts. In the Hadoop world, data is stored in varying forms. Path can help locate the dataset on a storage machine, but input format is what comes into play when actual reading takes place. To keep it short and sweet, input formats help generate splits from files which are logical representations of the data. Input formats provide record readers which do the actual reading from these splits.
+
+InputFormat interface exposes [2 methods](https://hadoop.apache.org/docs/stable/api/org/apache/hadoop/mapred/InputFormat.html) - getSplits() and getRecordReader().
+
+### Creation of splits
+*`getSplits()`* method facilitates the creation of splits from actual data files. On a high level, this involves reading the input paths from the Configuration object and scanning those paths to get the files present. Files are logically represented as FileStatus. These file paths are then passed to the `makeSplit()` method which creates the splits for further processing. Below are the snippets of the actual code where these calls are being made.
+
+![](/img/blog/2023-01-02-presto-hudi-deep-dive/get-splits-call.png)
+*Figure 1: listStatus() gets called internally from getSplits().*
+
+
+![](/img/blog/2023-01-02-presto-hudi-deep-dive/make-split-call.png)
+*Figure 2: makeSplit() call is highlighted as part of getSplits() implementation.*
+
+Above snippets are taken from FileInputFormat class. Please note that the lines are highlighted where these calls are being made. 
+
+## Hudi table types
+
+Hudi supports 2 types of tables out of the box, namely Copy on Write (CoW) and Merge on Read (MoR). CoW is the simpler of the two in terms of the reading and writing flow involved. Upserts made to this table type result in newer versions of columnar parquet files which can be read using the ParquetPageSource used internally by Presto.
+MoR involves writing incoming upserts to delta log files in avro format by default. These log files are later compacted to create new base parquet files. Since this involves row and columnar format files, this table type supports Read Optimized (RO) and real time (RT) queries.
+
+To be able to query these tables, the table metadata needs to be synced with Hive or glue metastore. Syncing MoR tables with HMS results in creation of 2 table types, `<table_name>_ro` and `<table_name>_rt`. They are internally differentiated by setting the config `hoodie.query.as.ro.table`. As the name suggests, the former is used for read optimized queries involving only base parquet files, while the latter is used for real time queries involving merging of delta log files and base parquet files on the fly.
+
+From Presto’s point of view, CoW table and `_ro` version of a MoR table behave the same way. Complexities occur when dealing with `_rt` version of MoR table type as we will see in the next section.
```

Comment:
Complexities ... ->
`_rt` version of MoR table type requires additional work as described in the next section.